File: backend-sqlite/src/controllers/importController.py
# backend-sqlite/src/controllers/importController.py
from fastapi import APIRouter, UploadFile, HTTPException, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
import pandas as pd
import io
import logging
from typing import List
from ..database import get_db
from .. import models, schemas

router = APIRouter()
logger = logging.getLogger(__name__)

@router.post("/api/import/productos")
async def importar_productos(file: UploadFile, db: Session = Depends(get_db)):
    try:
        # Leer el archivo
        contents = await file.read()
        
        # Determinar el tipo de archivo
        if file.filename.endswith(('.xlsx', '.xls')):
            df = pd.read_excel(io.BytesIO(contents))
        elif file.filename.endswith('.csv'):
            df = pd.read_csv(io.BytesIO(contents))
        else:
            raise HTTPException(status_code=400, detail="Formato de archivo no soportado")
        
        # Validar columnas mínimas requeridas
        required_columns = ['nombre', 'categoria', 'precio']
        for col in required_columns:
            if col not in df.columns:
                # Intentar mapear columnas similares
                matching_cols = [c for c in df.columns if col in str(c).lower()]
                if matching_cols:
                    df.rename(columns={matching_cols[0]: col}, inplace=True)
                else:
                    raise HTTPException(
                        status_code=400, 
                        detail=f"Columna requerida no encontrada: {col}"
                    )
        
        # Convertir a formato de salida
        productos = df.to_dict('records')
        productos_creados = []
        
        # Validar y guardar cada producto
        for producto_data in productos:
            try:
                # Validar y limpiar datos
                if 'precio' in producto_data:
                    try:
                        producto_data['precio'] = float(producto_data['precio'])
                    except (ValueError, TypeError):
                        producto_data['precio'] = 0.0
                
                # Asegurar que los campos requeridos tengan valores por defecto
                producto_data['activo'] = producto_data.get('activo', True)
                producto_data['stock'] = int(producto_data.get('stock', 0))
                
                # Verificar si el producto ya existe
                db_producto = db.query(models.Producto).filter(
                    models.Producto.nombre == str(producto_data['nombre'])
                ).first()
                
                if db_producto:
                    # Actualizar producto existente
                    for key, value in producto_data.items():
                        if hasattr(db_producto, key):
                            setattr(db_producto, key, value)
                else:
                    # Crear nuevo producto
                    db_producto = models.Producto(**{
                        'nombre': str(producto_data['nombre']),
                        'categoria': str(producto_data['categoria']),
                        'precio': float(producto_data['precio']),
                        'stock': int(producto_data.get('stock', 0)),
                        'descripcion': str(producto_data.get('descripcion', '')),
                        'codigo_barras': str(producto_data.get('codigo_barras', '')) if producto_data.get('codigo_barras') else None,
                        'activo': bool(producto_data.get('activo', True))
                    })
                    db.add(db_producto)
                
                db.commit()
                db.refresh(db_producto)
                productos_creados.append(db_producto)
                
            except Exception as e:
                db.rollback()
                logger.error(f"Error al procesar producto {producto_data.get('nombre')}: {str(e)}")
                continue
        
        return {
            "success": True,
            "message": f"Se importaron {len(productos_creados)} productos correctamente",
            "data": [{
                "id": p.id,
                "nombre": p.nombre,
                "categoria": p.categoria,
                "precio": p.precio,
                "stock": p.stock
            } for p in productos_creados]
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error al procesar archivo: {str(e)}")
        raise HTTPException(
            status_code=500, 
            detail=f"Error al procesar el archivo: {str(e)}"
        )

File: backend-sqlite/src/controllers/importController.js
import { exec } from 'child_process'
import { promisify } from 'util'
import path from 'path'
import { fileURLToPath } from 'url'
import fs from 'fs/promises'
import { existsSync } from 'fs'
import ProductoGeneral from '../models/ProductoGeneral.js'
import { respuestaExito } from '../utils/helpers.js'
import { AppError } from '../middlewares/errorHandler.js'

const execAsync = promisify(exec)
const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

/**
 * Procesa un archivo XLSX o PDF usando el script Python
 */
export const importarProductosDesdeArchivo = async (req, res) => {
  try {
    // Validar que se recibió un archivo
    if (!req.file) {
      // Validar también si el body está vacío o es null
      if (!req.body || Object.keys(req.body).length === 0) {
        return res.status(400).json({
          exito: false,
          mensaje: 'No se recibió ningún archivo. Asegúrese de enviar el archivo en el campo "archivo" con formato multipart/form-data'
        })
      }
      
      // Si hay body pero no file, puede ser que el campo no coincida
      return res.status(400).json({
        exito: false,
        mensaje: 'No se recibió ningún archivo. Verifique que el campo del formulario se llame "archivo"'
      })
    }

    const archivo = req.file
    const extension = path.extname(archivo.originalname).toLowerCase().slice(1)
    
    // Validar extensión
    if (!['xlsx', 'xls', 'pdf'].includes(extension)) {
      throw new AppError('Formato de archivo no soportado. Use XLSX, XLS o PDF', 400)
    }

    // Obtener API key de Gemini si está disponible (opcional para PDFs)
    const apiKey = req.body.apiKey || process.env.GEMINI_API_KEY || null

    // Ruta del script Python
    const scriptPath = path.join(__dirname, '../utils/importProducts.py')
    
    // Verificar que el script existe
    if (!existsSync(scriptPath)) {
      throw new AppError('Script de importación no encontrado', 500)
    }

    // Ejecutar script Python
    const comando = `python "${scriptPath}" ${extension} "${archivo.path}" ${apiKey ? `"${apiKey}"` : ''}`
    
    console.log('Ejecutando comando:', comando.replace(apiKey || '', '***'))
    
    let stdout
    let stderr = ''
    try {
      const result = await execAsync(comando, {
        maxBuffer: 10 * 1024 * 1024, // 10MB
        encoding: 'utf8'
      })
      stdout = result.stdout || ''
      stderr = result.stderr || ''
    } catch (execError) {
      // Limpiar archivo temporal en caso de error
      try {
        if (archivo.path && existsSync(archivo.path)) {
          await fs.unlink(archivo.path)
        }
      } catch (err) {
        console.error('Error al eliminar archivo temporal:', err)
      }
      
      // Capturar stdout y stderr del error
      const errorStdout = execError.stdout || ''
      const errorStderr = execError.stderr || ''
      
      // Intentar parsear stdout como JSON para obtener mensaje de error del script
      let errorMessage = 'Error al ejecutar el script de importación'
      if (errorStdout) {
        try {
          const errorData = JSON.parse(errorStdout.trim())
          if (errorData.mensaje) {
            errorMessage = errorData.mensaje
          } else if (errorData.error) {
            errorMessage = errorData.error
          }
        } catch {
          // Si no es JSON, usar el texto directo
          errorMessage = errorStdout.substring(0, 200) || errorMessage
        }
      } else if (errorStderr) {
        errorMessage = errorStderr.substring(0, 200)
      } else {
        errorMessage = execError.message || errorMessage
      }
      
      console.error('Error ejecutando script Python:', {
        message: errorMessage,
        stdout: errorStdout.substring(0, 200),
        stderr: errorStderr.substring(0, 200)
      })
      
      return res.status(500).json({
        exito: false,
        mensaje: errorMessage
      })
    }

    // Limpiar archivo temporal después de ejecutar
    try {
      if (archivo.path && existsSync(archivo.path)) {
        await fs.unlink(archivo.path)
      }
    } catch (err) {
      console.error('Error al eliminar archivo temporal:', err)
    }

    // Validar que stdout no esté vacío o sea "null"
    const stdoutTrimmed = (stdout || '').trim()
    
    // Verificar valores que indican error o respuesta inválida
    if (!stdoutTrimmed || 
        stdoutTrimmed === 'null' || 
        stdoutTrimmed.toLowerCase() === 'null' ||
        stdoutTrimmed === 'None' ||
        stdoutTrimmed.toLowerCase() === 'none') {
      console.error('Script Python devolvió respuesta vacía o inválida:', {
        stdout: stdoutTrimmed,
        stderr: stderr.substring(0, 500)
      })
      return res.status(400).json({
        exito: false,
        mensaje: 'El script de importación no devolvió datos válidos. Verifica que el archivo tenga el formato correcto y que Python esté instalado correctamente.'
      })
    }

    // Verificar que stdout parezca JSON válido (debe empezar con { o [)
    if (!stdoutTrimmed.startsWith('{') && !stdoutTrimmed.startsWith('[')) {
      console.error('Script Python devolvió respuesta que no es JSON:', {
        stdout: stdoutTrimmed.substring(0, 500),
        stderr: stderr.substring(0, 500)
      })
      return res.status(400).json({
        exito: false,
        mensaje: `El script de importación devolvió una respuesta inválida. Respuesta: ${stdoutTrimmed.substring(0, 200)}`
      })
    }

    // Parsear respuesta del script
    let resultado
    try {
      resultado = JSON.parse(stdoutTrimmed)
    } catch (parseError) {
      console.error('Error al parsear respuesta del script:', {
        error: parseError.message,
        stdout: stdoutTrimmed.substring(0, 500),
        stderr: stderr.substring(0, 500)
      })
      return res.status(500).json({
        exito: false,
        mensaje: `Error al procesar respuesta del script: ${parseError.message}. La respuesta recibida no es JSON válido.`
      })
    }

    // Verificar si hubo error en el script
    if (resultado.error) {
      return res.status(400).json({
        exito: false,
        mensaje: resultado.error
      })
    }

    // Verificar estructura de respuesta
    if (resultado.exito === false) {
      return res.status(400).json({
        exito: false,
        mensaje: resultado.mensaje || 'Error al procesar el archivo'
      })
    }

    if (!resultado.productos || !Array.isArray(resultado.productos)) {
      console.error('Formato de respuesta inválido:', resultado)
      return res.status(500).json({
        exito: false,
        mensaje: 'Formato de respuesta inválido del script. El script no devolvió la estructura esperada.'
      })
    }

    // Validar y crear productos en la base de datos
    const productosCreados = []
    const productosConError = []
    const usuarioId = req.usuario?.id || null

    for (const productoData of resultado.productos) {
      try {
        // Validar datos mínimos
        if (!productoData.nombre || productoData.nombre.trim() === '') {
          productosConError.push({
            producto: productoData,
            error: 'Nombre vacío'
          })
          continue
        }

        // Preparar datos del producto (mapear campos del script Python)
        const datosProducto = {
          nombre: productoData.nombre.trim(),
          codigoBarras: (productoData.codigoBarras && productoData.codigoBarras.trim() !== '') 
            ? productoData.codigoBarras.trim() 
            : null,
          costoBase: Number.parseFloat(productoData.costoBase || productoData.precio || 0) || 0,
          categoria: productoData.categoria || 'General',
          unidad: productoData.unidad || 'unidad',
          descripcion: productoData.descripcion || null,
          proveedor: productoData.proveedor || null,
          creadoPorId: usuarioId,
          tipoCreacion: 'importacion'
        }

        // Verificar si el producto ya existe (por código de barras o nombre)
        let productoExistente = null
        if (datosProducto.codigoBarras) {
          productoExistente = ProductoGeneral.buscarPorCodigoBarras(datosProducto.codigoBarras)
        }
        
        if (!productoExistente) {
          // Buscar por nombre (búsqueda aproximada)
          const productos = ProductoGeneral.buscar({
            buscar: datosProducto.nombre,
            limite: 1
          })
          
          if (productos.datos && productos.datos.length > 0) {
            const productoSimilar = productos.datos[0]
            // Si el nombre es muy similar, considerarlo duplicado
            if (productoSimilar.nombre.toLowerCase() === datosProducto.nombre.toLowerCase()) {
              productoExistente = productoSimilar
            }
          }
        }

        if (productoExistente) {
          // Actualizar producto existente
          ProductoGeneral.actualizar(productoExistente.id, {
            costoBase: datosProducto.costoBase,
            categoria: datosProducto.categoria
          })
          productosCreados.push({
            ...productoExistente,
            accion: 'actualizado'
          })
        } else {
          // Crear nuevo producto
          const nuevoProducto = ProductoGeneral.crear(datosProducto)
          productosCreados.push({
            ...nuevoProducto,
            accion: 'creado'
          })
        }

      } catch (error) {
        console.error('Error al procesar producto:', productoData, error)
        productosConError.push({
          producto: productoData,
          error: error.message || 'Error desconocido'
        })
      }
    }

    // Preparar respuesta
    const respuesta = {
      totalProcesados: resultado.productos.length,
      totalCreados: productosCreados.filter(p => p.accion === 'creado').length,
      totalActualizados: productosCreados.filter(p => p.accion === 'actualizado').length,
      totalErrores: productosConError.length,
      productos: productosCreados,
      errores: productosConError
    }

    res.json(respuestaExito(respuesta, `Importación completada: ${respuesta.totalCreados} creados, ${respuesta.totalActualizados} actualizados`))

  } catch (error) {
    // Limpiar archivo temporal en caso de error
    if (req.file?.path && existsSync(req.file.path)) {
      try {
        await fs.unlink(req.file.path)
      } catch (err) {
        console.error('Error al eliminar archivo temporal:', err)
      }
    }

    if (error instanceof AppError) {
      throw error
    }

    console.error('Error en importarProductosDesdeArchivo:', error)
    throw new AppError(`Error al importar productos: ${error.message}`, 500)
  }
}

export default {
  importarProductosDesdeArchivo
}

